internal/agent/device/fileio/writer.go

```diff
@@ -87,6 +88,35 @@ func (w *writer) MkdirAll(path string, perm os.FileMode) error {
 	return os.MkdirAll(filepath.Join(w.rootDir, path), perm)
 }
 
+func (w *writer) MkdirTemp(prefix string) (string, error) {
+	baseDir := w.rootDir
+	if baseDir == "" {
+		baseDir = os.TempDir()
+	}
+
+	for i := 0; i < 10000; i++ {
+		suffix := fmt.Sprintf("%06d", mrand.Intn(1000000)) // #nosec G404
+		name := prefix + suffix
+		path := filepath.Join(baseDir, name)
+
+		err := os.Mkdir(path, DefaultDirectoryPermissions)
+		if err != nil {
+			if !os.IsExist(err) {
+				return "", fmt.Errorf("create temp dir: %w", err)
+			}
+			// exists, try again
+			continue
+		}
+		if w.rootDir != "" {
+			// return the relative path
+			return name, nil
+		}
+		return path, nil
+	}
+
+	return "", fmt.Errorf("generate unique temp dir")
```

Comment:
Isn't this equivalent to
```suggestion
        path, err := os.MkdirTemp(baseDir, prefix)
        return strings.TrimPrefix(path, w.rootDir), err
```

Reply:
will double check but this looks good

Reply:
updated with tests